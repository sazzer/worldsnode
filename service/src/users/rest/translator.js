// @flow

import {type Resource} from '../../model/resource';
import {type UserData} from '../user';

/** The type representing a login the user has with a provider */
export type UserLoginModel = {
    '@type': string,
    'provider': string,
    'providerId': string,
    'providerName': string
};

/** The type representing the API Model of a User */
export type UserModel = {
    '@context': string,
    '@id': string,
    '@type': Array<string>,
    'created': Date,
    'updated': Date,
    'name': string,
    'email': ?string,
    'logins': Array<UserLoginModel>
};

/**
 * Actually translate the provided user to the model representation
 * @param user The user to translate
 * @return the API Model representation
 */
export function translateUserToModel(user: Resource<UserData>): UserModel {
    return {
        '@context': '/jsonld/users/User.jsonld',
        '@id': `/api/users/${user.identity.id}`,
        '@type': [
            'Person',
            'User',
        ],

        'created': user.identity.created,
        'updated': user.identity.updated,

        'name': user.data.name,
        'email': user.data.email,
        'logins': user.data.logins.map((provider) => ({
            '@type': 'Login',
            'provider': provider.provider,
            'providerId': provider.providerId,
            'providerName': provider.displayName,
        })),
    };
}
