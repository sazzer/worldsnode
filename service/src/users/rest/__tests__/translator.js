// @flow

import * as testSubject from '../translator';

describe('translateUserToModel', () => {
    test('Translating a minimal user', () => {
        const user = {
            identity: {
                id: '123',
                version: '1',
                created: new Date('2018-11-15T09:34:00Z'),
                updated: new Date('2018-11-15T09:34:00Z'),
            },
            data: {
                name: 'Graham',
                logins: [],
            },
        };
        const translated = testSubject.translateUserToModel(user);
        expect(translated).toMatchSnapshot();
    });

    test('Translating a full user', () => {
        const user = {
            identity: {
                id: '123',
                version: '1',
                created: new Date('2018-11-15T09:34:00Z'),
                updated: new Date('2018-11-15T09:34:00Z'),
            },
            data: {
                name: 'Graham',
                email: 'graham@grahamcox.co.uk',
                logins: [
                    {
                        provider: 'google',
                        providerId: '123456',
                        displayName: 'graham@grahamcox.co.uk',
                    },
                ],
            },
        };
        const translated = testSubject.translateUserToModel(user);
        expect(translated).toMatchSnapshot();
    });
});
