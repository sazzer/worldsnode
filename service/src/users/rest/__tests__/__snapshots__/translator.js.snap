// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateUserToModel Translating a full user 1`] = `
Object {
  "@context": "/jsonld/users/User.jsonld",
  "@id": "/api/users/123",
  "@type": Array [
    "Person",
    "User",
  ],
  "created": 2018-11-15T09:34:00.000Z,
  "email": "graham@grahamcox.co.uk",
  "logins": Array [
    Object {
      "@type": "Login",
      "provider": "google",
      "providerId": "123456",
      "providerName": "graham@grahamcox.co.uk",
    },
  ],
  "name": "Graham",
  "updated": 2018-11-15T09:34:00.000Z,
}
`;

exports[`translateUserToModel Translating a minimal user 1`] = `
Object {
  "@context": "/jsonld/users/User.jsonld",
  "@id": "/api/users/123",
  "@type": Array [
    "Person",
    "User",
  ],
  "created": 2018-11-15T09:34:00.000Z,
  "email": undefined,
  "logins": Array [],
  "name": "Graham",
  "updated": 2018-11-15T09:34:00.000Z,
}
`;
